data/pools/transactionPool.go

```diff
@@ -965,10 +974,50 @@ func (pool *TransactionPool) assembleEmptyBlock(round basics.Round) (assembled *
 		return nil, err
 	}
 	next := bookkeeping.MakeBlock(prev)
-	blockEval, err := pool.ledger.StartEvaluator(next.BlockHeader, 0)
+	blockEval, err := pool.ledger.StartEvaluator(next.BlockHeader, 0, pool.calculateMaxTxnBytesPerBlock(next.BlockHeader.CurrentProtocol))
 	if err != nil {
 		err = fmt.Errorf("TransactionPool.assembleEmptyBlock: cannot start evaluator for %d: %v", round, err)
 		return nil, err
 	}
 	return blockEval.GenerateBlock()
 }
+
+// SetDataExchangeRate updates the data exchange rate this node is expected to have.
+func (pool *TransactionPool) SetDataExchangeRate(dataExchangeRate uint64) {
+	atomic.StoreUint64(&pool.latestMeasuredDataExchangeRate, dataExchangeRate)
+}
+
+// calculateMaxTxnBytesPerBlock computes the optimal block size for the current node, based
+// on it's effective network capabilities. This number is bound by the protocol MaxTxnBytesPerBlock.
+func (pool *TransactionPool) calculateMaxTxnBytesPerBlock(consensusVersion protocol.ConsensusVersion) int {
```

Comment:
Thinking out loud for a moment... This introduces a new dependency on the protocol. While you didn't add the import, previously we were only using a `TxType` constant in this file. Maybe pass `dataExchangeRate` instead and do this computation in the evaluator?

Another option would be moving the function to Eval as a helper.

For testability, you can avoid the receiver here. The only dependencies are `config.ConsensusProtocols` and `pool.latestMeasuredDataExchangeRate`, if you pass them directly there is no concern about side effects.

Reply:
starting from the end:
I think that for testability, it would be very trivial as is ( I'll add few tests ).
It's definitely doable for the evaluator to "consume" that value directly off the transaction pool.. but it makes it much more linked - which isn't a good thing. Having this computation being done on either the transaction pool or the evaluator is fine - but I don't know if we need the evaluator to be aware of "network" related constructs.

In fact, I think it would have been fine moving this logic from the transaction pool into the transaction sync connector if it wasn't for the block-specific version, which might affect the block size.